### PR TITLE
introduce step action based structs

### PR DESCRIPTION
**Pull request by geoberle**
### What this PR does

* added `Step` minimal interface for step metadata
* introduce `ShellStep` and `ARMStep` structs for existing steps to hold action specific parameters
* introduce new step types for DNS child zone delegation and KV issuer configuration

Jira: 
Link to demo recording:

### CODE REVIEW

**Review by janboll — APPROVED** [review]
:rocket:

File: tooling/templatize/pkg/pipeline/types.go [hunk 1]
```diff
@@ -14,25 +16,170 @@ type Pipeline struct {
...
+	for i, rawStep := range rawRg.Steps {
+		switch rawStep.meta.Action {
+		case "Shell":
+			rg.Steps[i] = &ShellStep{}
+		case "ARM":
+			rg.Steps[i] = &ARMStep{}
+		default:
+			return fmt.Errorf("unknown action type %s", rawStep.meta.Action)
+		}
+		err := rawStep.unmarshal(rg.Steps[i])
```
Comment (janboll) [inline tooling/templatize/pkg/pipeline/types.go:48]: don't you have to call UnmarshalYAML?

Comment (geoberle) [reply tooling/templatize/pkg/pipeline/types.go:48]: `UnmarshalYAML` this is a "magic" function the golang yaml library calls

File: tooling/templatize/pkg/pipeline/run_test.go [hunk 2]
```diff
@@ -9,82 +9,54 @@ import (
 
 func TestStepRun(t *testing.T) {
 	fooundOutput := ""
```
Comment (stevekuznetsov) [inline tooling/templatize/pkg/pipeline/run_test.go:11]:
```suggestion
	foundOutput := ""
```

File: tooling/templatize/pkg/pipeline/run_test.go [hunk 3]
```diff
@@ -9,82 +9,54 @@ import (
...
+	)
+	_, err := RunStep(s, context.Background(), "", &executionTargetImpl{}, &PipelineRunOptions{}, nil)
 	assert.NilError(t, err)
 	assert.Equal(t, fooundOutput, "hello\n")
 }
 
-func TestStepRunSkip(t *testing.T) {
-	s := &Step{
-		Name: "step",
+/*func TestStepRunSkip(t *testing.T) {
```
Comment (stevekuznetsov) [inline tooling/templatize/pkg/pipeline/run_test.go:22]: Prefer to delete, comments get stale.

File: tooling/templatize/pkg/pipeline/types.go [hunk 4]
```diff
@@ -14,25 +16,170 @@ type Pipeline struct {
...
+	}
+	msg.unmarshal = unmarshal
+	return nil
+}
+
+func (msg *rawStep) Unmarshal(v interface{}) error {
+	return msg.unmarshal(v)
+}
+
+type Step interface {
```
Comment (stevekuznetsov) [inline tooling/templatize/pkg/pipeline/types.go:94]: No `Run()`?

File: tooling/templatize/pkg/pipeline/types.go [hunk 5]
```diff
@@ -14,25 +16,170 @@ type Pipeline struct {
...
+		Command: command,
+	}
+}
+
+type ShellStep struct {
+	StepMeta   `yaml:",inline"`
+	Command    string     `yaml:"command,omitempty"`
+	Variables  []Variable `yaml:"variables,omitempty"`
+	DryRun     DryRun     `yaml:"dryRun,omitempty"`
+	outputFunc outPutHandler
```
Comment (stevekuznetsov) [inline tooling/templatize/pkg/pipeline/types.go:116]: This smells a tiny bit - and makes static validation of your code harder by storing function pointers with data, which renders things like `cmp.Diff()` useless. Who creates the output handler, is it a runtime concern? Why does it need to be attached to the step as opposed to somehow passed in to a `Run()`?

File: tooling/templatize/pkg/pipeline/types.go (hunk 4 above)
Comment (geoberle) [reply tooling/templatize/pkg/pipeline/types.go:94]: i had a `Run` in there to begin with but then reconsidered. i have high hopes that these structures will be the ones we will harmonize onto when we merge the code from ARO-HCP + ADO-Pipelines into the new repo. and for the EV2 context, a run function does not make any sense. so  i tried to keep them as data-only as possible.

File: tooling/templatize/pkg/pipeline/run_test.go (hunk 3 above)
Comment (geoberle) [reply tooling/templatize/pkg/pipeline/run_test.go]: gone

File: tooling/templatize/pkg/pipeline/run_test.go (hunk 2 above)
Comment (geoberle) [reply tooling/templatize/pkg/pipeline/run_test.go]: ty

File: tooling/templatize/pkg/pipeline/types.go (hunk 5 above)
Comment (geoberle) [reply tooling/templatize/pkg/pipeline/types.go:116]: i agree. it has been here already so no change. let's tackle this in a followup PR. it should be part of the Run func somehow

File: tooling/templatize/pkg/pipeline/types.go [hunk 6]
```diff
@@ -14,25 +16,170 @@ type Pipeline struct {
...
+func (m *StepMeta) Dependencies() []string {
+	return m.DependsOn
+}
+
+type rawStep struct {
+	meta      *StepMeta
+	unmarshal func(interface{}) error
+}
+
+func (msg *rawStep) UnmarshalYAML(unmarshal func(interface{}) error) error {
```
[the remark this answers is not in the document]
Comment (geoberle) [reply tooling/templatize/pkg/pipeline/types.go:81]: this is a way to remember the actual full content of a step, stored as the unmarshal func. this way we can turn a `rawStep` into `XXXStep` later. 

do you propose another option?

[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply tooling/templatize/pkg/pipeline/types.go:81]: The uninteresting approach would be e.g. https://goplay.tools/snippet/yBhVozOXrvB - up to you, IMO a little cleaner as you are in control of the codepath explicitly as opposed to implicitly having UnmarshalYAML called and storing the func for later. Little easier to wrap my brain around.

**Pull request merged**